File: aether-backend/scripts/validate_job_types.py
#!/usr/bin/env python3
# Incoming: none --- {none, none}
# Processing: none --- {0 jobs: none}
# Outgoing: none --- {none, none}
"""
Job Type Registry Validator

Validates that all @.architecture documentation uses only registered job types.

@.architecture
Incoming: .architecture/data_flow_standard.yaml, Backend .py files with @.architecture --- {Dict[str, Any], utf-8}
Processing: parse_registry(), extract_used_jobs(), validate_compliance(), report_violations() --- {3 jobs: JOB_LOG, JOB_TRACE, JOB_VALIDATE_SCHEMA}
Outgoing: stdout report, process exit code --- {Dict[str, Any], cli}
"""

import re
import sys
import yaml
from pathlib import Path
from typing import Dict, Set, List, Tuple, Any
from collections import defaultdict


def _strip_architecture_header(raw: str) -> str:
    """Remove three-line @.architecture docstring header from YAML text."""
    parts = raw.split("\n\n", 1)
    return parts[1] if len(parts) == 2 else raw


def _load_registry_data(registry_path: Path) -> Dict[str, Any]:
    """Resolve registry data, following source_of_truth pointers if declared."""
    raw_content = _strip_architecture_header(registry_path.read_text(encoding='utf-8'))
    raw = yaml.safe_load(raw_content)

    if not isinstance(raw, dict):
        raise RuntimeError(f"Job registry at {registry_path} must deserialize to a mapping.")

    source = raw.get('source_of_truth')
    if isinstance(source, str):
        source_path = Path(source)
        if not source_path.is_absolute():
            source_path = (registry_path.parent / source_path).resolve()
        source_content = _strip_architecture_header(source_path.read_text(encoding='utf-8'))
        raw = yaml.safe_load(source_content)
        if not isinstance(raw, dict):
            raise RuntimeError(f"Job registry source {source_path} must deserialize to a mapping.")

    return raw


def load_job_registry(registry_path: Path) -> Set[str]:
    """Load all registered job types from data_flow_standard.yaml"""
    raw_content = registry_path.read_text(encoding='utf-8')
    
    # Remove @.architecture header if present
    raw_content = _strip_architecture_header(raw_content)
    
    registry = yaml.safe_load(raw_content)
    
    if not isinstance(registry, dict):
        raise RuntimeError(f"Job registry at {registry_path} must deserialize to a mapping.")

    job_types = set()

    # Load from job_types array
    job_type_list = registry.get('job_types', [])
    if isinstance(job_type_list, list):
        for job_entry in job_type_list:
            if isinstance(job_entry, dict) and 'name' in job_entry:
                job_types.add(job_entry['name'])
            elif isinstance(job_entry, str):
                job_types.add(job_entry)
    
    # Also check for legacy 'categories' structure
    categories = registry.get('categories', {})
    if isinstance(categories, dict):
        for category_data in categories.values():
            jobs = category_data.get('jobs', [])
            for job_name in jobs:
                job_types.add(job_name)

    return job_types


def extract_jobs_from_file(file_path: Path) -> List[str]:
    """Extract job types from a file's @.architecture documentation"""
    try:
        content = file_path.read_text(encoding='utf-8')
        
        # Find Processing line in @.architecture and capture the job list payload
        arch_match = re.search(
            r'@\.architecture.*?Processing:\s*[^\n]+?---\s*\{([^}]+)\}',
            content,
            re.DOTALL
        )
        
        if not arch_match:
            return []
        
        jobs_str = arch_match.group(1).strip()
        jobs_str = re.sub(r'^\d+\s+jobs?:\s*', '', jobs_str, flags=re.IGNORECASE)
        
        # Split by comma and clean
        jobs = [j.strip() for j in jobs_str.split(',') if j.strip()]
        
        return jobs
        
    except Exception as e:
        print(f"⚠️  Error reading {file_path}: {e}")
        return []


def find_python_files(root_dir: Path, exclude_dirs: List[str]) -> List[Path]:
    """Find all Python files excluding specified directories"""
    files = []
    for py_file in root_dir.rglob("*.py"):
        if any(excluded in py_file.parts for excluded in exclude_dirs):
            continue
        if py_file.name == "__init__.py":
            continue
        files.append(py_file)
    return files


def find_architecture_yaml_files(backend_root: Path) -> List[Path]:
    """Locate backend architecture YAML manifests/directives."""
    architecture_dir = backend_root / ".architecture"
    if architecture_dir.exists():
        files = {
            path
            for pattern in ("*.yaml", "*.yml")
            for path in architecture_dir.rglob(pattern)
        }
        return sorted(files)
    return []


def extract_jobs_from_yaml(file_path: Path) -> List[str]:
    """Extract job identifiers from backend architecture YAML."""
    try:
        raw_text = file_path.read_text(encoding='utf-8')
    except Exception as exc:
        print(f"⚠️  Error reading YAML {file_path}: {exc}")
        return []

    header_section = raw_text.split("\n\n", 1)[0]
    header_jobs = re.findall(r'JOB_[A-Z0-9_]{2,}', header_section)

    try:
        payload = _strip_architecture_header(raw_text)
        data = yaml.safe_load(payload)
    except yaml.YAMLError as exc:
        print(f"⚠️  Error parsing YAML {file_path}: {exc}")
        data = None

    jobs: Set[str] = set()

    def _collect(node: Any) -> None:
        if isinstance(node, dict):
            for value in node.values():
                _collect(value)
        elif isinstance(node, list):
            for item in node:
                _collect(item)
        elif isinstance(node, str):
            matches = re.findall(r'JOB_[A-Z0-9_]{2,}', node)
            jobs.update(matches)

    if data is not None:
        _collect(data)

    if not jobs:
        jobs.update(re.findall(r'JOB_[A-Z0-9_]{2,}', raw_text))
    else:
        jobs.update(re.findall(r'JOB_[A-Z0-9_]{2,}', raw_text))

    jobs.update(header_jobs)

    return sorted(job.strip() for job in jobs if job.strip())


def validate_job_types(backend_root: Path) -> Tuple[bool, Dict]:
    """Validate all files use only registered job types"""
    
    # Load registry from .architecture/data_flow_standard.yaml
    registry_path = backend_root / ".architecture" / "data_flow_standard.yaml"
    if not registry_path.exists():
        print(f"❌ Registry not found: {registry_path}")
        print(f"   Expected location: {backend_root}/.architecture/data_flow_standard.yaml")
        return False, {}
    
    registered_jobs = load_job_registry(registry_path)
    print(f"📋 Loaded {len(registered_jobs)} registered job types from {registry_path.relative_to(backend_root)}")
    
    # Find all Python files
    exclude_dirs = ['services', 'venv', '__pycache__', 'tests', '.cache', 'node_modules']
    py_files = find_python_files(backend_root, exclude_dirs)
    print(f"🔍 Scanning {len(py_files)} Python files...\n")
    
    # Track violations
    violations: Dict[str, Set[str]] = defaultdict(set)
    python_files_with_arch = 0
    unregistered_jobs = set()
    
    for py_file in py_files:
        jobs = extract_jobs_from_file(py_file)
        if not jobs:
            continue
        
        python_files_with_arch += 1
        rel_path = py_file.relative_to(backend_root)
        
        for job in jobs:
            if job not in registered_jobs:
                violations[job].add(str(rel_path))
                unregistered_jobs.add(job)

    yaml_files = find_architecture_yaml_files(backend_root)
    print(f"📑 Scanning {len(yaml_files)} architecture YAML files...\n")

    yaml_files_with_jobs = 0

    for yaml_file in yaml_files:
        jobs = extract_jobs_from_yaml(yaml_file)
        if not jobs:
            continue

        yaml_files_with_jobs += 1
        rel_path = yaml_file.relative_to(backend_root)

        for job in jobs:
            if job not in registered_jobs:
                violations[job].add(str(rel_path))
                unregistered_jobs.add(job)
    
    # Report results
    print("=" * 80)
    print("📊 VALIDATION RESULTS")
    print("=" * 80)
    print(f"\n✅ Python files scanned: {len(py_files)}")
    print(f"✅ Python files with @.architecture: {python_files_with_arch}")
    print(f"✅ Architecture YAML files scanned: {len(yaml_files)}")
    print(f"✅ Architecture YAML files with jobs: {yaml_files_with_jobs}")
    print(f"✅ Registered job types: {len(registered_jobs)}")
    
    if violations:
        print(f"\n❌ Unregistered job types found: {len(unregistered_jobs)}\n")
        
        for job in sorted(unregistered_jobs):
            paths = sorted(violations[job])
            print(f"\n⚠️  Unregistered job type: '{job}'")
            print(f"   Used in {len(paths)} file(s):")
            for file_path in paths[:5]:  # Show first 5
                print(f"   - {file_path}")
            if len(paths) > 5:
                print(f"   ... and {len(paths) - 5} more")
        
        print("\n" + "=" * 80)
        print("❌ VALIDATION FAILED")
        print("=" * 80)
        print("\nTo fix:")
        print("1. Add missing job types to .architecture/data_flow_standard.yaml")
        print("2. Or update files to use registered job types")
        print("3. Run this script again to verify")
        
        return False, violations
    else:
        print("\n✅ All job types are registered!\n")
        print("=" * 80)
        print("✅ VALIDATION PASSED")
        print("=" * 80)
        return True, {}


def main():
    """Main entry point"""
    print("\n🔍 Backend Job Type Registry Validator\n")
    
    # Find backend root
    script_dir = Path(__file__).parent
    backend_root = script_dir.parent
    
    # Validate
    valid, violations = validate_job_types(backend_root)
    
    # Exit with appropriate code
    sys.exit(0 if valid else 1)


if __name__ == '__main__':
    main()

File: aether-frontend/scripts/validate_job_types.js
#!/usr/bin/env node
'use strict';

/**
 * @.architecture
 * 
 * Incoming: Frontend .js/.jsx/.ts/.tsx files with @.architecture --- {JavaScript source files, regex patterns}
 * Processing: extractUsedJobs(), validateTokenFormat(), (optional) validateRegistryMembership() --- {4 jobs: JOB_PARSE_JSON, JOB_ROUTE_BY_TYPE, JOB_VALIDATE_SCHEMA, JOB_EMIT_EVENT}
 * Outgoing: stdout (validation report), exit code (0=valid, 1=invalid) --- {string console output, number exit code}
 * 
 * Job Type Validator
 * ==================================================
 * Validates that @.architecture documentation uses valid JOB_* tokens.
 * Optional: enforce registry membership via --strict-registry.
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

function stripArchitectureHeader(raw) {
  if (typeof raw !== 'string') {
    return raw;
  }
  const headerPattern = /^Incoming:[^\n]*\nProcessing:[^\n]*\nOutgoing:[^\n]*\n*/;
  return raw.replace(headerPattern, '');
}

// ============================================================================
// Load Registry (Optional)
// ============================================================================

function loadJobRegistry(registryPath) {
  const content = fs.readFileSync(registryPath, 'utf8');
  const registry = yaml.load(stripArchitectureHeader(content));

  const jobTypes = new Set();

  const catalog = registry && typeof registry === 'object' ? registry.catalog : undefined;
  if (catalog && typeof catalog === 'object') {
    for (const categoryValue of Object.values(catalog)) {
      if (!categoryValue || typeof categoryValue !== 'object') {
        continue;
      }

      const entries = Array.isArray(categoryValue.entries) ? categoryValue.entries : [];
      for (const entry of entries) {
        const id = entry && typeof entry === 'object' ? entry.id : undefined;
        if (typeof id === 'string' && id.trim().length > 0) {
          jobTypes.add(id.trim());
        }
      }
    }
  }

  const categories = registry && typeof registry === 'object' ? registry.categories : undefined;
  if (categories && typeof categories === 'object') {
    for (const categoryValue of Object.values(categories)) {
      if (!categoryValue || typeof categoryValue !== 'object') {
        continue;
      }

      const jobs = Array.isArray(categoryValue.jobs) ? categoryValue.jobs : [];
      for (const job of jobs) {
        if (typeof job === 'string' && job.trim().length > 0) {
          jobTypes.add(job.trim());
        }
      }
    }
  }

  return jobTypes;
}

// ============================================================================
// Extract Jobs from Files
// ============================================================================

function normalizeJobToken(token) {
  if (typeof token !== 'string') return null;
  const t = token.trim();
  if (!t) return null;
  return t;
}

function isValidJobToken(job) {
  return /^JOB_[A-Z0-9_]{2,}$/.test(job);
}

function extractJobsFromFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    
    // Find Processing line in @.architecture
    // Format: Processing: ... --- {N jobs: JOB_X, JOB_Y, ...}
    const archRegex = /@\.architecture.*?Processing:.*?---\s*\{[^}]*jobs?:\s*([^}]+)\}/s;
    const match = content.match(archRegex);
    
    if (!match) {
      return { jobs: [], invalid: [] };
    }
    
    const jobsStr = match[1];
    
    // Split by comma and clean
    const jobs = jobsStr
      .split(',')
      .map(normalizeJobToken)
      .filter(Boolean);

    const invalid = jobs.filter(job => job !== 'none' && !isValidJobToken(job));
    return { jobs, invalid };
    
  } catch (err) {
    console.error(`⚠️  Error reading ${filePath}: ${err.message}`);
    return { jobs: [], invalid: ['__read_error__'] };
  }
}

// ============================================================================
// Find Files
// ============================================================================

async function findJavaScriptFiles(rootDir, excludeDirs) {
  const files = [];
  
  async function walk(currentPath) {
    const entries = await fs.promises.readdir(currentPath, { withFileTypes: true });
    
    for (const entry of entries) {
      const fullPath = path.join(currentPath, entry.name);
      
      if (entry.isDirectory()) {
        if (!excludeDirs.includes(entry.name) && !entry.name.startsWith('.')) {
          await walk(fullPath);
        }
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name);
        if (['.js', '.jsx', '.ts', '.tsx'].includes(ext)) {
          files.push(fullPath);
        }
      }
    }
  }
  
  await walk(rootDir);
  return files;
}

// ============================================================================
// Validate
// ============================================================================

async function validateJobTypes(frontendRoot) {
  const strictRegistry = process.argv.includes('--strict-registry');

  // Load registry (optional)
  const registryPath = path.join(frontendRoot, '..', 'Architecture', 'frontend_job_registry.yaml');
  const hasRegistry = fs.existsSync(registryPath);
  if (strictRegistry && !hasRegistry) {
    console.error(`Registry not found (strict): ${registryPath}`);
    return { valid: false, violations: { __registry_missing__: [registryPath] } };
  }

  const registeredJobs = hasRegistry ? loadJobRegistry(registryPath) : null;
  if (registeredJobs) {
    console.log(`Loaded ${registeredJobs.size} registered job types from registry`);
  } else {
    console.log('Registry: not configured (skipping membership enforcement). Use --strict-registry to require it.');
  }
  
  // Find all JS/TS files
  const excludeDirs = ['node_modules', '.next', 'dist', 'build', 'coverage', 'test-results'];
  const jsFiles = await findJavaScriptFiles(frontendRoot, excludeDirs);
  console.log(`Scanning ${jsFiles.length} JavaScript/TypeScript files...\n`);
  
  // Track violations
  const violations = {};
  const invalidJobTokens = {};
  let filesWithArch = 0;
  const unregisteredJobs = new Set();
  
  for (const jsFile of jsFiles) {
    const { jobs, invalid } = extractJobsFromFile(jsFile);
    if (jobs.length === 0 && invalid.length === 0) {
      continue;
    }
    
    filesWithArch++;
    const relPath = path.relative(frontendRoot, jsFile);

    if (invalid.length > 0) {
      invalidJobTokens[relPath] = invalid;
    }
    
    for (const job of jobs) {
      if (job === 'none') {
        continue;
      }
      if (registeredJobs && !registeredJobs.has(job)) {
        if (!violations[job]) {
          violations[job] = [];
        }
        violations[job].push(relPath);
        unregisteredJobs.add(job);
      }
    }
  }
  
  // Report results
  console.log('='.repeat(80));
  console.log('📊 VALIDATION RESULTS');
  console.log('='.repeat(80));
  console.log(`\n✅ Files scanned: ${jsFiles.length}`);
  console.log(`✅ Files with @.architecture: ${filesWithArch}`);
  if (registeredJobs) {
    console.log(`✅ Registered job types: ${registeredJobs.size}`);
  }

  const invalidFiles = Object.keys(invalidJobTokens).length;
  if (invalidFiles > 0) {
    console.log(`\n❌ Invalid job tokens found in ${invalidFiles} file(s)\n`);
    Object.entries(invalidJobTokens)
      .slice(0, 10)
      .forEach(([file, toks]) => {
        console.log(`⚠️  ${file}`);
        console.log(`   Invalid: ${toks.join(', ')}`);
      });
    if (invalidFiles > 10) {
      console.log(`... and ${invalidFiles - 10} more`);
    }
  }

  if (registeredJobs && Object.keys(violations).length > 0) {
    console.log(`\n❌ Unregistered job types found: ${unregisteredJobs.size}\n`);
    
    const sortedJobs = Array.from(unregisteredJobs).sort();
    for (const job of sortedJobs) {
      console.log(`\n⚠️  Unregistered job type: '${job}'`);
      console.log(`   Used in ${violations[job].length} file(s):`);
      const filesToShow = violations[job].slice(0, 5);
      for (const filePath of filesToShow) {
        console.log(`   - ${filePath}`);
      }
      if (violations[job].length > 5) {
        console.log(`   ... and ${violations[job].length - 5} more`);
      }
    }
    
    console.log('\n' + '='.repeat(80));
    console.log('❌ VALIDATION FAILED');
    console.log('='.repeat(80));
    
    return { valid: false, violations };
  }

  if (invalidFiles > 0) {
    console.log('\n' + '='.repeat(80));
    console.log('❌ VALIDATION FAILED');
    console.log('='.repeat(80));
    return { valid: false, violations: { __invalid_job_tokens__: invalidJobTokens } };
  }

  {
    console.log(`\n✅ All job types are registered!\n`);
    console.log('='.repeat(80));
    console.log('✅ VALIDATION PASSED');
    console.log('='.repeat(80));
    return { valid: true, violations: {} };
  }
}

// ============================================================================
// Main
// ============================================================================

async function main() {
  console.log('\nFrontend Job Type Validator\n');
  
  // Find frontend root
  const scriptDir = __dirname;
  const frontendRoot = path.dirname(scriptDir);
  
  // Validate
  const result = await validateJobTypes(frontendRoot);
  
  // Exit with appropriate code
  process.exit(result.valid ? 0 : 1);
}

// Run
main().catch(err => {
  console.error('❌ Fatal error:', err);
  process.exit(1);
});

